rcl/test/rcl/test_publisher_wait_all_ack.cpp

```diff
@@ -0,0 +1,204 @@
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  publisher_options.qos.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+  publisher_options.qos.depth = 10000;
+  ret = rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  rcl_subscription_options_t subscription_options = rcl_subscription_get_default_options();
+  subscription_options.qos.depth = 1;
+  subscription_options.qos.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+
+  INIT_SUBSCRIPTION(1)
+  INIT_SUBSCRIPTION(2)
+  INIT_SUBSCRIPTION(3)
+
+  ASSERT_TRUE(wait_for_established_subscription(&publisher, 10, 100));
+
+  char test_string[ONE_MEGABYTE];
+  memset(test_string, 'a', ONE_MEGABYTE);
+  test_string[ONE_MEGABYTE - 1] = '\0';
+  test_msgs__msg__Strings msg;
+  test_msgs__msg__Strings__init(&msg);
+  ASSERT_TRUE(rosidl_runtime_c__String__assign(&msg.string_value, test_string));
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    test_msgs__msg__Strings__fini(&msg);
+  });
+
+  ret = rcl_publish(&publisher, &msg, nullptr);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription1, context_ptr, 10, 100));
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription2, context_ptr, 10, 100));
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription3, context_ptr, 10, 100));
+
+  int i = 0;
+  for (; i < 500; i++) {
+    ret = rcl_publish(&publisher, &msg, nullptr);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+
+  ret = rcl_publisher_wait_for_all_acked(
+    &publisher,
+    RCL_MS_TO_NS(500));
+  EXPECT_TRUE(ret == RCL_RET_OK || ret == RCL_RET_TIMEOUT);
+
+  ret = rcl_publisher_wait_for_all_acked(&publisher, -1);
+  EXPECT_EQ(RCL_RET_OK, ret);
+}
+
+TEST_F(
+  CLASSNAME(TestPublisherFixtureSpecial, RMW_IMPLEMENTATION),
+  test_wait_for_all_acked_with_best_effort)
+{
+  rcl_ret_t ret;
+  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes);
+  constexpr char topic_name[] = "test_wait_for_all_acked_with_best_effort";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  publisher_options.qos.reliability = RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT;
+  publisher_options.qos.depth = 10000;
+  ret = rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
```

Comment:
EXPECT_EQ() inside a OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(), doesn't work, because it's just returning from the lambda (see [here](https://github.com/google/googletest/blob/master/docs/advanced.md#propagating-fatal-failures)).

You can use the [FAIL()](https://github.com/google/googletest/blob/master/docs/reference/assertions.md#fail-fail) macro instead.